src/hardware/mouse/mouse.cpp

```diff
@@ -0,0 +1,652 @@
+		queue.AddEvent(ev);
+	}
+}
+
+void MOUSE_EventWheel(const int16_t w_rel)
+{
+	MouseEvent ev;
+	for (auto &interface : mouse_interfaces)
+		if (interface->IsUsingHostPointer())
+			interface->NotifyWheel(ev, w_rel);
+	queue.AddEvent(ev);
+}
+
+void MOUSE_EventWheel(const int16_t w_rel, const MouseInterfaceId interface_id)
+{
+	auto interface = MouseInterface::Get(interface_id);
+	if (interface && interface->IsUsingEvents()) {
+		MouseEvent ev;
+		interface->NotifyWheel(ev, w_rel);
+		queue.AddEvent(ev);
+	}
+}
+
+// ***************************************************************************
+// MOUSECTL.COM / GUI configurator interface
+// ***************************************************************************
+
+static std::vector<MouseInterface *> get_relevant_interfaces(
+        const std::vector<MouseInterfaceId> &list_ids)
+{
+	std::vector<MouseInterface *> list_tmp = {};
+
+	if (list_ids.empty())
+		// If command does not specify interfaces,
+		// assume we are interested in all of them
+		list_tmp = mouse_interfaces;
+	else
+		for (const auto &interface_id : list_ids) {
+			auto interface = MouseInterface::Get(interface_id);
+			if (interface)
+				list_tmp.push_back(interface);
+		}
+
+	// Filter out not emulated ones
+	std::vector<MouseInterface *> list_out = {};
+	for (const auto &interface : list_tmp)
+		if (interface->IsEmulated())
+			list_out.push_back(interface);
+
+	return list_out;
+}
+
+MouseControlAPI::MouseControlAPI()
+{
+	manymouse.StartConfigAPI();
+}
+
+MouseControlAPI::~MouseControlAPI()
+{
+	manymouse.StopConfigAPI();
+	MOUSE_NotifyStateChanged();
+}
+
+bool MouseControlAPI::IsNoMouseMode()
+{
+	return mouse_config.no_mouse;
+}
+
+const std::vector<MouseInterfaceInfoEntry> &MouseControlAPI::GetInfoInterfaces() const
+{
+	return mouse_info.interfaces;
+}
+
+const std::vector<MousePhysicalInfoEntry> &MouseControlAPI::GetInfoPhysical()
+{
+	manymouse.RescanIfSafe();
+	return mouse_info.physical;
+}
+
+bool MouseControlAPI::CheckInterfaces(const MouseControlAPI::ListIDs &list_ids)
+{
+	const auto list = get_relevant_interfaces(list_ids);
+
+	if (list_ids.empty() && list.empty())
+		return false; // no emulated mouse interfaces
+	if (list_ids.empty())
+		return true; // OK, requested all emulated interfaces
+	if (list_ids.size() != list.size())
+		return false; // at least one requested interface is not emulated
+
+	return true;
+}
+
+bool MouseControlAPI::PatternToRegex(const std::string &pattern, std::regex &regex)
+{
+	// Convert DOS wildcard pattern to a regular expression
+	std::stringstream pattern_regex;
+	pattern_regex << std::hex;
+	for (const auto character : pattern) {
+		if (character < 0x20 || character > 0x7E) {
+			return false;
+		}
+		if (character == '?')
+			pattern_regex << ".";
+		else if (character == '*')
+			pattern_regex << ".*";
+		else if ((character >= '0' && character <= '9') ||
+		         (character >= 'a' && character <= 'z') ||
+		         (character >= 'A' && character <= 'Z'))
```

Comment:
is alpha-numeric: `int isalnum( int ch );` from `<cctype>` maybe can do the job? 

https://en.cppreference.com/w/cpp/string/byte/isalnum

Reply:
Changed.